runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java

```diff
@@ -169,18 +172,28 @@ private boolean isLaneComplete(Path laneDir, IlluminaNotificationDto dto) {
   public NotificationDto process(File runDirectory, TimeZone tz) throws IOException {
     // Call the C++ program to do the real work and write a notification DTO to standard output. The C++ object has no direct binding to the
     // DTO, so any changes to the DTO must be manually changed in the C++ code.
-    ProcessBuilder builder = new ProcessBuilder("nice", "runscanner-illumina", runDirectory.getAbsolutePath()).directory(runDirectory);
+    ProcessBuilder builder = new ProcessBuilder("nice", "runscanner-illumina", runDirectory.getAbsolutePath()).directory(runDirectory)
+        .redirectError(Redirect.INHERIT);
     builder.environment().put("TZ", tz.getID());
     Process process = builder.start();
 
-    IlluminaNotificationDto dto = createObjectMapper().readValue(process.getInputStream(), IlluminaNotificationDto.class);
-    dto.setSequencerFolderPath(runDirectory.getAbsolutePath());
-    try {
-      if (process.waitFor() != 0) {
-        throw new IOException("Illumina run processor did not exit cleanly: " + runDirectory.getAbsolutePath());
+    IlluminaNotificationDto dto;
+    int exitcode;
+    try (InputStream output = process.getInputStream(); OutputStream input = process.getOutputStream()) {
+      dto = createObjectMapper().readValue(output, IlluminaNotificationDto.class);
+      dto.setSequencerFolderPath(runDirectory.getAbsolutePath());
+    } finally {
+      try {
+        exitcode = process.waitFor();
+      } catch (InterruptedException e) {
+        throw new IOException(e);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not throw exceptions in finally blocks. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1163)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this throw statement from this finally block. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1143)